File: app/src/main/java/com/example/furiganakeyboard/ime/RomajiConversionState.kt
package com.example.furiganakeyboard.ime

/**
 * State transitions for Japanese IME conversion keys on the romaji panel.
 *
 * Candidate presentation is deliberately not a state transition: scrolling the
 * candidate bar must not turn a pre-conversion composition into a selection.
 */
class RomajiConversionState {
    enum class Phase { EMPTY, COMPOSING, SELECTING_CANDIDATE }

    sealed interface KeyAction {
        data object InsertSpace : KeyAction
        data object SendEditorAction : KeyAction
        data object CommitComposition : KeyAction
        data class SelectCandidate(val index: Int) : KeyAction
        data class CommitCandidate(val index: Int) : KeyAction
    }

    var phase: Phase = Phase.EMPTY
        private set

    private var selectedIndex = 0

    /** Call only when romaji input has been edited, not when candidates redraw. */
    fun onCompositionEdited(hasComposition: Boolean) {
        phase = if (hasComposition) Phase.COMPOSING else Phase.EMPTY
        selectedIndex = 0
    }

    fun clear() = onCompositionEdited(hasComposition = false)

    fun selectedCandidateIndex(candidateCount: Int): Int? =
        if (phase == Phase.SELECTING_CANDIDATE && candidateCount > 0) {
            selectedIndex.mod(candidateCount)
        } else {
            null
        }

    fun onSpace(candidateCount: Int): KeyAction = when (phase) {
        Phase.EMPTY -> KeyAction.InsertSpace
        Phase.COMPOSING -> selectFirstOrCommit(candidateCount)
        Phase.SELECTING_CANDIDATE -> selectNextOrCommit(candidateCount)
    }

    fun onEnter(candidateCount: Int): KeyAction = when (phase) {
        Phase.EMPTY -> KeyAction.SendEditorAction
        Phase.COMPOSING -> commitFirstOrComposition(candidateCount)
        Phase.SELECTING_CANDIDATE -> commitSelectedOrComposition(candidateCount)
    }

    private fun selectFirstOrCommit(candidateCount: Int): KeyAction {
        if (candidateCount <= 0) return KeyAction.CommitComposition
        phase = Phase.SELECTING_CANDIDATE
        selectedIndex = 0
        return KeyAction.SelectCandidate(selectedIndex)
    }

    private fun selectNextOrCommit(candidateCount: Int): KeyAction {
        if (candidateCount <= 0) return KeyAction.CommitComposition
        selectedIndex = (selectedIndex + 1).mod(candidateCount)
        return KeyAction.SelectCandidate(selectedIndex)
    }

    private fun commitFirstOrComposition(candidateCount: Int): KeyAction =
        if (candidateCount > 0) KeyAction.CommitCandidate(0) else KeyAction.CommitComposition

    private fun commitSelectedOrComposition(candidateCount: Int): KeyAction =
        selectedCandidateIndex(candidateCount)?.let(KeyAction::CommitCandidate)
            ?: KeyAction.CommitComposition
}

File: ios/Sources/KeyboardExtension/RomajiConversionState.swift
import Foundation

/// Conversion-key state shared by Space and Return on the Japanese romaji panel.
final class RomajiConversionState {
    enum Phase: Equatable {
        case empty
        case composing
        case selectingCandidate
    }

    enum Action: Equatable {
        case insertSpace
        case sendEditorAction
        case commitComposition
        case selectCandidate(Int)
        case commitCandidate(Int)
    }

    private(set) var phase: Phase = .empty
    private var selectedIndex = 0

    func compositionEdited(hasComposition: Bool) {
        phase = hasComposition ? .composing : .empty
        selectedIndex = 0
    }

    func clear() {
        compositionEdited(hasComposition: false)
    }

    func selectedCandidateIndex(candidateCount: Int) -> Int? {
        guard phase == .selectingCandidate, candidateCount > 0 else { return nil }
        return selectedIndex % candidateCount
    }

    func space(candidateCount: Int) -> Action {
        switch phase {
        case .empty:
            return .insertSpace
        case .composing:
            guard candidateCount > 0 else { return .commitComposition }
            phase = .selectingCandidate
            selectedIndex = 0
            return .selectCandidate(0)
        case .selectingCandidate:
            guard candidateCount > 0 else { return .commitComposition }
            selectedIndex = (selectedIndex + 1) % candidateCount
            return .selectCandidate(selectedIndex)
        }
    }

    func enter(candidateCount: Int) -> Action {
        switch phase {
        case .empty:
            return .sendEditorAction
        case .composing:
            return candidateCount > 0 ? .commitCandidate(0) : .commitComposition
        case .selectingCandidate:
            guard let index = selectedCandidateIndex(candidateCount: candidateCount) else {
                return .commitComposition
            }
            return .commitCandidate(index)
        }
    }
}
